src/global.cpp

```diff
@@ -36,27 +36,82 @@ int symPACK_Rank(int * rank){
 }
 
 
+#ifdef CUDA_MODE
+extern "C"
+void compute_device_alloc_size() {
+  int n_gpus = upcxx::gpu_default_device::device_n();
+  if (n_gpus==0) {
+    std::cerr<<"Error: found no CUDA devices"<<std::endl;
+    abort();
+  }
+
+  MPI_Comm shmcomm;
+  int tasks_per_node;
+  MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0, MPI_INFO_NULL, &shmcomm);
+  MPI_Comm_size(shmcomm, &tasks_per_node);
+  
+  size_t alloc_size, free, total;
+  CUDA_ERROR_CHECK(cudaMemGetInfo(&free, &total));
+  alloc_size = (free/2) / (std::max(tasks_per_node, n_gpus) / n_gpus);
```

Comment:
This algorithm assumes that `n_gpus == cuda_device::device_n()` reflects the _total_ number of physical GPUs plugged into this compute node, which is often false.

SLURM GPU allocation options can and often do cause CUDA_VISIBLE_DEVICES to be set in a way such that `libcuda` (and hence also `cuda_device`) for each task only sees the fraction of physical GPUs assigned to that task. For example this command:
```
srun --gpus-per-node=4 -N 1 -n 4 a.out
```
results in 4 processes sharing a node that each "see" all **four** shared GPUs (as reported by `cuda_device::device_n()` and `cuDeviceGetCount()`), and the algorithm should work as intended.
However this command:
```
srun --gpus-per-task=1 -N 1 -n 4 a.out   
```
results in 4 processes sharing a node that each only see their **single** dedicated GPU, and this algorithm will underestimate the space available to each by a factor of four.
Similarly this command:
```
srun --ntasks-per-gpu=16 -N 1 -n 64 a.out
```
results in 64 processes sharing a node that each only see the **single** GPU shared with 15 peers, and this algorithm will also underestimate the space available to each by a factor of 4.

TL:DR: On systems with more than 1 physical GPU per node, some job layouts can cause this defaulting algorithm to choose a size considerably smaller than an even partitioning of the shared GPU memory resources. Please add a comment noting this deficiency and update README to explain the defaulting for `-gpu_mem` might not work optimally for systems with more than 1 GPU/node, where manual setting should be recommended.

Reply:
Thanks for adding documentation in the README.

I still think a comment in the code (and/or an enhancement item in TODO.md) is also warranted to record the fact this defaulting algorithm is known to greatly underestimate the usable space for some layouts on multi-GPU systems.